### DISCUSSION TITLE
m5 Stack Core2 native capacitive touch buttons

### DISCUSSION THREAD

**Discussion by kairos0ne** [Moddable-OpenSource/moddable#1132] [General]
I was wondering what module to use to interact with the core2 cap-touch buttons. 

I have built a menu system following the example on the wifi-config example in the docs. 

I guess I could go with pins/digital module and setup a loop but I've seen examples where device is imported from device.

And I've also seen examples where m5button is used and I'm not sure whats the best approach for this. 

Anyone got any advice.

**Comment by phoddie** [comment]
When you build for the M5Stack Core2 using `-p esp32/m5stack_core2`, support for the built-in capacitive touch is included. You can poll the buttons.

```js
trace(button.a.read(), "\n");
trace(button.b.read(), "\n");
trace(button.c.read(), "\n");
```

Or, to eliminate polling, you can install a callback function to be invoked when the button changes state.

```js
button.a.onChanged = function() {
   trace(this.read(), "\n");
}
```

> I guess I could go with pins/digital module and setup a loop but I've seen examples where device is imported from device.

That works for many devices, but not here. The buttons on M5Stack Core2 are virtual. They are actually part of the touch screen, so they aren't directly connected to a GPIO.

**Comment by mkellner** [comment]
Looking in `$MODDABLE/build/devices/esp32/targets/m5stack_core2` I see there is a `M5StackCoreTouch.js` file with a class that extends `Touch`. In that class, it writes a 1 or 0 to global button a, b, or c.

In `setup-target.js` global buttons `a`, `b`, `c` are set up as M5Core2Buttons.

If button.a has an `onChanged` function, it will be called when it is touched or un-touched.

So, for the `esp32/m5stack_core2` platform, I think you just have to set the global `button.a.onChanged = function () { your function }`

**Comment by kairos0ne** [comment]
Thanks, that works well. 

This is fantastic platform, I'm very impressed with what you have achieved with it.

**Reply by phoddie** [reply]
Thank you!! Glad that worked for you.
